### ISSUE TITLE
ResNet giving different outputs in TF and TFLite in C++

### ISSUE THREAD

**Issue by suraj-maniyar** [tensorflow/tensorflow#50274]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux CentOS 8
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.2
- Python version: 3.6.8
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): 8.3.1

**Describe the current behavior**
ResNet36 (and ResNet50) models give different outputs when running inference on TensorFlow and TensorFlow-Lite in C++.

**Describe the expected behavior**
Outputs should be same across all platforms and across different APIs.

**Standalone code to reproduce the issue**
Due to certain restrictions, I won't be able to provide fully reproducible code. Following are some snippets : 
Model : 
```
base_model = tf.keras.applications.ResNet50(
    include_top=False, weights='imagenet',
    input_shape=[96,112,3])

x = base_model.output
x = tf.keras.layers.Flatten()(x)
x = tf.keras.layers.Dense(128, use_bias=True, name='Bottleneck')(x)
   
model = tf.keras.models.Model( base_model.input, x )
```

TFLite inference : 
```
std::unique_ptr<tflite::FlatBufferModel> flat_buffer_model = tflite::FlatBufferModel::BuildFromBuffer((const char*)model_buffer->data(), model_buffer->size(), &err_reporter);
tflite::ops::builtin::BuiltinOpResolver builtins;
std::unique_ptr<tflite::Interpreter> interpreter;
tflite::InterpreterBuilder(*flat_buffer_model, builtins)(&interpreter);

const std::vector<int>& outputs = thread_interpreter.interpreter_->outputs();
TfLiteStatus tf_stat = thread_interpreter.interpreter_->Invoke();

```

**Other info / logs** 
I have a ResNet36 model trained using TFv1 and saved into a protobuf file format. This is used to generate inference on sample image using TensorFlow C API on windows. 
I used ```tf.compat.v1.lite.converter.from_frozen_graph()``` to convert this protobuf to a .tflite file.
I ran inference from TFLite using C++ API (on x86_64 architecture).
Additionally, I also ran inference from the TFLite model using Python.

```***All the 3 outputs are different.***```

Interestingly, I also have an InceptionResNet model (.pb file from TF1 converted into .tflite file using TF2) and it gives the exact same output when I run inference in TF (C) and TFLite (C++). So, this makes me question that something is wrong with tflite converter/API specifically when dealing with ResNet architecture.

Note : I also tested ResNet50 architecture which was trained in TF2 and got similar anomalous results. (This is just to rule out any functionalities that may not have been supported by backwards compatible converter).

I am interested in knowing pointers such as gcc compiler optimizations or anything else that may cause different outputs across TensroFlow and TensorFlow-Lite in C++.

Any help is much appreciated.

**Comment by suraj-maniyar** [comment]
@saikumarchalla Let me know if you need any more information.

**Comment by jdduke** [comment]
Hi @suraj-maniyar , can you include the code/command you used to convert the model? In the past, the main source of different outputs in vision models is differences in the preprocessing pipeline for the image (how it is scaled and/or cropped), so if you want to include the representative code you used that would be helpful. You might also try running the TFLite Interpreter from Python with the same input to see if you get different results.

**Comment by suraj-maniyar** [comment]
@jdduke 

I used this code to convert .pb to .tflite : 
```

import tensorflow as tf
tf.__version__
# 2.3.2

graph_def_file = "resface36_L2Norm128_at_epoch_168.pb"
converter = tf.compat.v1.lite.TFLiteConverter.from_frozen_graph(graph_def_file,
                                              ["input"],
                                              ["resface36/Bottleneck/BiasAdd"],
                                              input_shapes={'input': [1, 96, 112, 3]})
tflite_model = converter.convert()

open("resface36_L2Norm128_at_epoch_168.tflite", "wb").write(tflite_model)

```


As for the preprocessing pipeline, I am not even using any preprocessing on the image. I simply created a matrix of zeros of dimension 96x112 and fed it to network in TF and TFLite. I have also tested Inception ResNet and it gave exact same output in TF and TFLite. I only have problems with ResNet where the outputs don't match.

And yes, I do get different output when running TFLite interpreter from Python. 

So basically, I have 3 different outputs for the same resnet model with the same input (zero matrix) on different configurations (TF on C++, TFLite on C++, TFLite on Python)

**Comment by srjoglekar246** [comment]
Is the output correct when using TFLite via Python? Or does TFLite provide wrong (but different) outputs via C++ & Python? It would be helpful to see the TFLite model (untrained is fine) you have, so I can take a look.

Also I assume you are doing `interpreter->AllocateTensors()` when using C++, before any invoke?

**Comment by pkgoogle** [comment]
Hi @suraj-maniyar, can you give us a better idea on how different is different? what scale are we talking about here? These aren't usually direct 1:1 conversions, some optimizations are applied and during the whole process some precision is lost, so any additional information will be great!
